### PR TITLE
Add Unix peer credentials authenticator

### PR DESCRIPTION
**Pull request by joechrisellis**
This authenticator uses Unix peer credentials for authentication. Unix
peer credentials provide direct access to the (effective) uid/gid on the
other end of a domain socket connect, without cooperation between the
endpoints. This means that we can determine the uid/gid of the
connecting process, and therefore infer the username of the user that
owns said process.

This authenticator:

- grabs the (uid, gid) pair of the connecting process.
- determine the username of the owner of the connecting process based on
  the uid.
- creates an `ApplicationName` based on the username.

Note that this patch depends on the following PR being merged:

    https://github.com/rust-lang/rust/pull/75148

At the time of writing, this PR is currently under review and is not
merged into the Rust stdlib. This patch therefore will not build with
the current a stable/nightly compiler.

Signed-off-by: Joe Ellis <joe.ellis@arm.com>

### CODE REVIEW

**Comment by joechrisellis** [comment]
Note: CI failure is expected since this patch depends on a feature that is not yet in the stdlib.

File: src/authenticators/uds_authenticator/mod.rs
Comment (paulhowardarm) [inline src/authenticators/uds_authenticator/mod.rs:7]: Do we gain anything by converting to a user name as the application name, rather than just using the stringified UID itself as the application ID string? This seems to have additional dependencies and operations that could conceivably fail. I'm also slightly fearful about converting to usernames given that the environment may be containerised, which could have implications in terms of how the UIDs convert to names in the transition between containers.

File: src/authenticators/mod.rs
Comment (paulhowardarm) [inline src/authenticators/mod.rs:15]: Is it useful to perhaps wrap this authenticator in a Rust/Cargo feature so that it can be selected for at build time? (Actually, looking at the rest of the change, perhaps you've already done that? But there's an indication that it's only temporary. I actually think there is value in having this be a selectable thing even longer term, because it's useful for people to be able to compile down a Parsec executable with only the precise pieces that their deployment needs. Providers and authenticators are the kinds of things that we might want to snip out, because they can be heavyweight, especially in terms of the dependencies that get dragged in with them).

File: src/authenticators/mod.rs
Comment (paulhowardarm) [inline src/authenticators/mod.rs:12]: This comment is outdated by the change.

File: src/authenticators/uds_authenticator/mod.rs
Comment (paulhowardarm) [inline src/authenticators/uds_authenticator/mod.rs:1]: I guess this should be 2020 now.

File: src/authenticators/uds_authenticator/mod.rs
Comment (hug-dev) [reply src/authenticators/uds_authenticator/mod.rs:7]: > Do we gain anything by converting to a user name as the application name, rather than just using the stringified UID itself as the application ID string?

I think the only thing we gain is for it to look nicer in the `mappings` folder and in logs like `"New request received from application name \"{}\""`. So mostly for esthetics/debugging. Also, if we add any operation which takes an application name as parameter, it might be easier to specify the user name rather than the UID.

I guess I supposed that every UID must have an username mapping to it but if that is not true in some environments, I agree using the plain UID would be better.

Comment (joechrisellis) [reply src/authenticators/uds_authenticator/mod.rs:7]: That's a good point.

I can't think of any reason why the stringified UID wouldn't work just fine. The only comment that I have is that the username is perhaps more 'human-friendly', but I don't think these strings will ever necessarily be exposed anyways.

Happy to change this. 🙂

File: src/authenticators/mod.rs
Comment (joechrisellis) [reply src/authenticators/mod.rs:15]: Yes, good idea. Will do this. 👍

File: src/authenticators/mod.rs
Comment (joechrisellis) [reply src/authenticators/mod.rs:12]: Ah, well spotted, thanks!

File: src/authenticators/uds_authenticator/mod.rs
Comment (joechrisellis) [reply src/authenticators/uds_authenticator/mod.rs:1]: Fixed. :)

File: src/authenticators/uds_authenticator/mod.rs
Comment (ionut-arm) [inline src/authenticators/uds_authenticator/mod.rs:39]: I thought the authentication was meant to verify that the `RequestAuth` contains a string with that UID, just as a "sanity" check

Comment (joechrisellis) [reply src/authenticators/uds_authenticator/mod.rs:39]: Ah, I wasn't aware. Will fix that. 👍

Comment (paulhowardarm) [reply src/authenticators/uds_authenticator/mod.rs:39]: Agreed, I think it should. I didn't follow through every line here and I thought there was some Rusty magic going on. My opinion is that this system should be fully explicit with a well-defined authentication type and expected header contents. The calling process should "declare" its UID in the auth header, and the service should use the peer credential to check that it matches.

Comment (joechrisellis) [reply src/authenticators/uds_authenticator/mod.rs]: Yeah, it's definitely a fair point. This is now implemented. 🙂

File: src/authenticators/mod.rs
Comment (paulhowardarm) [inline src/authenticators/mod.rs:15]: I should have noticed this before, but I'm not sure why this is called UDS authenticator? Assuming UDS stands for Unix Domain Socket, that's the name of the transport rather than the method of authentication (which is UID). You could have all sorts of different authenticators on a UDS transport, so calling this a UDS authenticator doesn't quite characterise it accurately.

Comment (ionut-arm) [reply src/authenticators/mod.rs:15]: Does `PeerCredentialAuthenticator` make more sense? I can imagine we'd be able to get peer credentials for other IPC mechanisms as well and we could apply the same logic for authentication there

Comment (joechrisellis) [reply src/authenticators/mod.rs:15]: My bad; I made the wrong assumption that using a Unix domain socket as a transport meant that we would always want to use peer credentials for authentication. I agree with @ionut-arm, `PeerCredentialsAuthenticator` seems like a good name here. 🙂

File: Cargo.toml
Comment (hug-dev) [inline Cargo.toml:73]: See [my related comment](https://github.com/parallaxsecond/parsec-book/pull/57#discussion_r485089884), if it is agreed, I would personally prefer all occurences of "peer credentials" to be renamed "Unix peer credentials".

File: src/front/domain_socket.rs
Comment (hug-dev) [inline src/front/domain_socket.rs:206]: Is the `GetMetadata` trait needed here if we are already in Unix specific code? As in, could you just replace the implementation of that trait on `UnixStream` here?

File: src/front/domain_socket.rs
Comment (hug-dev) [inline src/front/domain_socket.rs:257]: I would prefer returning an error instead of panicking. In our case, you can log something and return `None`!

**Review by hug-dev — COMMENTED** [review]
Thanks for adding it! Probably better to wait for the specs on the book and the PR in Rust to be merged before merging this one but thought of some early comments!

File: src/authenticators/unix_peer_credentials_authenticator/mod.rs
Comment (hug-dev) [inline src/authenticators/unix_peer_credentials_authenticator/mod.rs:65]: I think you can directly use the `TryInto` from the slice to the owned type and use the failed case as a way to check that the auth buffer was not 4 bytes (so even checks if not empty 😃 ):
```rust
let expected_uid = auth.buffer.expose_secret();
let expected_uid: [u8; 4] = expected_uid_bytes.try_into().ok_or_else(|| {
    error!(
        "UID in authentication request is not the right size (expected: {}, got: {}).",
        EXPECTED_UID_SIZE_BYTES,
        expected_uid_bytes.len()
    );
    return Err(ResponseStatus::AuthenticationError);
})?;
let expected_uid = u32::from_le_bytes(expected_uid);
```

Comment (joechrisellis) [reply src/authenticators/unix_peer_credentials_authenticator/mod.rs:65]: Ooh, that's much nicer. 😄

File: src/authenticators/unix_peer_credentials_authenticator/mod.rs
Comment (hug-dev) [inline src/authenticators/unix_peer_credentials_authenticator/mod.rs:104]: Pretty good idea to use that for unit tests!

File: src/front/domain_socket.rs
Comment (hug-dev) [reply src/front/domain_socket.rs:206]: Just pinging this question as I think it is still valid

File: src/front/domain_socket.rs
Comment (hug-dev) [inline src/front/domain_socket.rs:274]: Could you also please add the stabilizing issue https://github.com/rust-lang/rust/issues/42839?

**Review by hug-dev — COMMENTED** [review]
Good idea of having that code directly here 👌

File: src/authenticators/unix_peer_credentials_authenticator/mod.rs
Comment (joechrisellis) [reply src/authenticators/unix_peer_credentials_authenticator/mod.rs:52]: > Small nit but I think that the empty case will also be checked in the following try_into (as the buffer will not be 4 bytes).

Nice catch, done.

> I would also say that you can use the hardcoded 4 in the following code instead of EXPECTED_UID_SIZE_BYTES as this size os probably not something meant to change often.

I don't know -- I kind of like this. It gives some obvious context to the magic number; when the user reads it, it is clear that the UID is supposed to be 4 bytes long. If that variable didn't exist it would be as clear, I don't think. 🙂

File: src/front/domain_socket.rs
Comment (joechrisellis) [reply src/front/domain_socket.rs:206]: We could -- this was more just thinking for the future. Will revert this, apologies for the lag on responding to this comment. 😄

File: src/front/domain_socket.rs
Comment (joechrisellis) [reply src/front/domain_socket.rs:266]: Done.

**Review by hug-dev — APPROVED** [review]
Thanks for applying the changes, it looks pretty good to me 👌

File: src/authenticators/unix_peer_credentials_authenticator/mod.rs
Comment (hug-dev) [reply src/authenticators/unix_peer_credentials_authenticator/mod.rs]: > I don't know -- I kind of like this. It gives some obvious context to the magic number; when the user reads it, it is clear that the UID is supposed to be 4 bytes long. If that variable didn't exist it would be as clear, I don't think. 🙂

Yes, it's true that this is more readable!

File: src/front/listener.rs
Comment (ionut-arm) [inline src/front/listener.rs:43]: Would it make sense to add the `pid` here as well? Don't know what it would be used for (if at all), but then again, we don't use the `gid` either...

File: src/authenticators/unix_peer_credentials_authenticator/mod.rs
Comment (ionut-arm) [inline src/authenticators/unix_peer_credentials_authenticator/mod.rs:187]: I'd say an issue in github or even better a non-documentation comment on `ConnectionMetadata` are more likely to catch a developer's attention if some other variant is added to the `enum`, otherwise this will probs just be forgotten

File: src/authenticators/unix_peer_credentials_authenticator/mod.rs
Comment (ionut-arm) [inline src/authenticators/unix_peer_credentials_authenticator/mod.rs:69]: Can you add the wildcard pattern and "disable" the clippy warning? Or is it something the compiler doesn't like?

File: src/authenticators/unix_peer_credentials_authenticator/mod.rs
Comment (joechrisellis) [reply src/authenticators/unix_peer_credentials_authenticator/mod.rs:187]: Added a comment in the `ConnectionMetadata` struct that asks the user to grep for `TODO(new_metadata_variant)` in the tests and update accordingly when a new variant is added.

File: src/front/listener.rs
Comment (joechrisellis) [reply src/front/listener.rs:43]: True, good for completeness -- added.

File: src/authenticators/unix_peer_credentials_authenticator/mod.rs
Comment (joechrisellis) [reply src/authenticators/unix_peer_credentials_authenticator/mod.rs]: We can -- done.

**Review by ionut-arm — APPROVED** [review]
Thank you! :100:

**Pull request merged**
